File: satellites/branch-snapshots/codex-jeeves-control-plane-v2/skeleton/kernel/governance.py
"""Governance — charters, edicts, and the decide gate (charter law).

Port of Apeloff1/gameforge-rs ``crates/gf-gameforge/src/governance.rs``.

This is charter/edict law for consequential domains — NOT agent utility
(``agents/policy.py``), NOT organism conductor policy, NOT cognition gates.
Every consequential mutate/submit crosses ``decide``; unchartered domains
fail closed.
"""

from __future__ import annotations

import threading
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Dict, List, Optional


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


@dataclass(frozen=True)
class Rule:
    """One written rule inside a domain charter."""

    id: str
    action: str
    min_weight: int = 0
    requires_quorum: bool = False


@dataclass
class Charter:
    id: str
    domain: str
    rules: List[Rule] = field(default_factory=list)
    ratified: datetime = field(default_factory=_utcnow)
    amendments: int = 0


@dataclass
class Edict:
    id: str
    charter_id: str
    rule: Rule
    proposed_by: str
    proposed_at: datetime = field(default_factory=_utcnow)
    in_force: bool = False


@dataclass(frozen=True)
class Decision:
    permitted: bool
    cited_rule: Optional[str]
    reason: str


class Governance:
    """Charter-law registry. Thread-safe; fail-closed on missing charters."""

    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._charters: Dict[str, Charter] = {}
        self._edicts: Dict[str, Edict] = {}

    def ratify(self, domain: str, rules: List[Rule]) -> Charter:
        """Ratify a charter for a domain (founding law needs no precedent)."""
        charter = Charter(
            id=str(uuid.uuid4()),
            domain=domain,
            rules=list(rules),
            ratified=_utcnow(),
            amendments=0,
        )
        with self._lock:
            self._charters[domain] = charter
        return charter

    def propose_edict(
        self, charter_domain: str, rule: Rule, proposed_by: str
    ) -> Optional[str]:
        """Propose an amendment held until the court enforces it."""
        with self._lock:
            charter = self._charters.get(charter_domain)
            if charter is None:
                return None
            edict = Edict(
                id=str(uuid.uuid4()),
                charter_id=charter.id,
                rule=rule,
                proposed_by=proposed_by,
                proposed_at=_utcnow(),
                in_force=False,
            )
            self._edicts[edict.id] = edict
            return f"edict:{edict.id}"

    def enforce_edict(self, edict_id: str) -> Optional[bool]:
        """Bring an edict into force after the court decided it."""
        with self._lock:
            edict = self._edicts.get(edict_id)
            if edict is None:
                return None
            if edict.in_force:
                return True
            edict.in_force = True
            charter_id = edict.charter_id
            rule = edict.rule
            for charter in self._charters.values():
                if charter.id == charter_id:
                    charter.rules.append(rule)
                    charter.amendments += 1
                    break
            return True

    def decide(self, domain: str, action: str, actor_weight: int) -> Decision:
        """Charter-law gate. Unchartered domains permit nothing."""
        with self._lock:
            charter = self._charters.get(domain)
            if charter is None:
                return Decision(
                    permitted=False,
                    cited_rule=None,
                    reason=f"no charter ratified for domain '{domain}'",
                )
            match = next((r for r in charter.rules if r.action == action), None)
            if match is None:
                return Decision(
                    permitted=False,
                    cited_rule=None,
                    reason=f"action '{action}' not written into the {domain} charter",
                )
            if actor_weight < match.min_weight:
                return Decision(
                    permitted=False,
                    cited_rule=match.id,
                    reason=(
                        f"actor weight {actor_weight} below required {match.min_weight}"
                    ),
                )
            reason = (
                "permitted by charter; quorum still required"
                if match.requires_quorum
                else "permitted by charter"
            )
            return Decision(permitted=True, cited_rule=match.id, reason=reason)

    def charters(self) -> List[Charter]:
        with self._lock:
            return list(self._charters.values())

    def edicts(self) -> List[Edict]:
        with self._lock:
            return list(self._edicts.values())

    def has_charter(self, domain: str) -> bool:
        with self._lock:
            return domain in self._charters

    def clear(self) -> None:
        """Test helper: wipe all charters/edicts."""
        with self._lock:
            self._charters.clear()
            self._edicts.clear()


# --- process singleton + founding law for forge / swarm mutate routes ---

_FOUNDING_FORGE: List[Rule] = [
    Rule(id="forge.blueprint", action="blueprint", min_weight=0),
    Rule(id="forge.materialise", action="materialise", min_weight=0),
    Rule(id="forge.archetype", action="archetype", min_weight=0),
    Rule(id="forge.run", action="run", min_weight=0),
    Rule(id="forge.intake", action="intake", min_weight=0),
]

_FOUNDING_SWARM: List[Rule] = [
    Rule(id="swarm.submit", action="submit", min_weight=0),
]

_gov: Optional[Governance] = None
_gov_lock = threading.Lock()


def reset_governance(gov: Optional[Governance] = None) -> Governance:
    """Replace the process singleton (tests). None → fresh empty Governance."""
    global _gov
    with _gov_lock:
        _gov = gov if gov is not None else Governance()
        return _gov


def ensure_founding_charters(gov: Governance) -> None:
    """Ratify forge/swarm founding law when missing (idempotent)."""
    if not gov.has_charter("forge"):
        gov.ratify("forge", list(_FOUNDING_FORGE))
    if not gov.has_charter("swarm"):
        gov.ratify("swarm", list(_FOUNDING_SWARM))


def get_governance(*, bootstrap: bool = True) -> Governance:
    """Process-wide governance. Bootstraps founding forge/swarm charters."""
    global _gov
    with _gov_lock:
        if _gov is None:
            _gov = Governance()
        if bootstrap:
            ensure_founding_charters(_gov)
        return _gov

File: satellites/gameforge-rs/crates/gf-gameforge/src/governance.rs
use std::collections::HashMap;
use chrono::{DateTime,Utc};
use serde::{Deserialize,Serialize};
use tokio::sync::RwLock;
use uuid::Uuid;

#[derive(Debug,Clone,Serialize,Deserialize)] pub struct Rule{pub id:String,pub action:String,pub min_weight:u32,pub requires_quorum:bool}
#[derive(Debug,Clone,Serialize,Deserialize)] pub struct Charter{pub id:String,pub domain:String,pub rules:Vec<Rule>,pub ratified:DateTime<Utc>,pub amendments:u32}
#[derive(Debug,Clone,Serialize,Deserialize)] pub struct Edict{pub id:String,pub charter_id:String,pub rule:Rule,pub proposed_by:String,pub proposed_at:DateTime<Utc>,pub in_force:bool}
#[derive(Debug,Clone,Serialize,Deserialize)] pub struct Decision{pub permitted:bool,pub cited_rule:Option<String>,pub reason:String}
pub struct Governance{charters:RwLock<HashMap<String,Charter>>,edicts:RwLock<HashMap<String,Edict>>}
impl Default for Governance{fn default()->Self{Self::new()}}
impl Governance{
 pub fn new()->Self{Self{charters:RwLock::new(HashMap::new()),edicts:RwLock::new(HashMap::new())}}
 pub async fn ratify(&self,domain:&str,rules:Vec<Rule>)->Charter{let c=Charter{id:Uuid::new_v4().to_string(),domain:domain.to_string(),rules,ratified:Utc::now(),amendments:0};self.charters.write().await.insert(c.domain.clone(),c.clone());c}
 pub async fn propose_edict(&self,domain:&str,rule:Rule,by:&str)->Option<String>{let c=self.charters.read().await;let charter=c.get(domain)?;let e=Edict{id:Uuid::new_v4().to_string(),charter_id:charter.id.clone(),rule,proposed_by:by.to_string(),proposed_at:Utc::now(),in_force:false};let p=format!("edict:{}",e.id);self.edicts.write().await.insert(e.id.clone(),e);Some(p)}
 pub async fn enforce_edict(&self,id:&str)->Option<()>{let mut es=self.edicts.write().await;let e=es.get_mut(id)?;if e.in_force{return Some(())}e.in_force=true;let cid=e.charter_id.clone();let rule=e.rule.clone();drop(es);let mut cs=self.charters.write().await;for c in cs.values_mut(){if c.id==cid{c.rules.push(rule);c.amendments+=1;break}}Some(())}
 pub async fn decide(&self,domain:&str,action:&str,actor_weight:u32)->Decision{let cs=self.charters.read().await;let Some(c)=cs.get(domain) else{return Decision{permitted:false,cited_rule:None,reason:format!("no charter ratified for domain '{domain}'")}};match c.rules.iter().find(|r|r.action==action){Some(r) if actor_weight>=r.min_weight=>Decision{permitted:true,cited_rule:Some(r.id.clone()),reason:if r.requires_quorum{"permitted by charter; quorum still required".into()}else{"permitted by charter".into()}},Some(r)=>Decision{permitted:false,cited_rule:Some(r.id.clone()),reason:format!("actor weight {actor_weight} below required {}",r.min_weight)},None=>Decision{permitted:false,cited_rule:None,reason:format!("action '{action}' not written into the {domain} charter")}}}
 pub async fn charters(&self)->Vec<Charter>{self.charters.read().await.values().cloned().collect()}
 pub async fn edicts(&self)->Vec<Edict>{self.edicts.read().await.values().cloned().collect()}
}
